examples/boot_counter/flash_sector.rs

```diff
@@ -0,0 +1,91 @@
+use core::{
+    mem::{size_of, MaybeUninit},
+    ptr::read_volatile,
+};
+
+use rp2040_flash::flash;
+
+/// XIP base address (see `XIP_BASE` in RP2040 datasheet).
+pub const FLASH_ORIGIN: usize = 0x10000000;
+/// The erasable sector size.
+const FLASH_SECTOR_SIZE: usize = 4096;
+/// The value an erased sector is filled with. This is typically 0xff.
+const FLASH_ERASED_VALUE: u8 = 0xff;
+
+/// The payload type `T` must fit into a single flash sector.
+///
+/// The payload type should be `repr(C)` to have a stable layout,
+/// because the flash-stored payload can survive firmware upgrades.
+///
+/// This data type itself should be flash sector aligned.
+pub union FlashSector<T>
+where
+    T: Copy,
+{
+    data: [u8; FLASH_SECTOR_SIZE],
+    value: MaybeUninit<T>,
+}
+
+impl<T> Default for FlashSector<T>
+where
+    T: Copy,
+{
+    fn default() -> Self {
+        Self {
+            data: [FLASH_ERASED_VALUE; FLASH_SECTOR_SIZE],
+        }
+    }
+}
+
+impl<T> FlashSector<T>
+where
+    T: Copy,
+{
+    pub fn new(value: T) -> Self {
+        assert!(
+            size_of::<T>() <= FLASH_SECTOR_SIZE,
+            "`T` must fit into a single sector size"
+        );
+        let mut instance = Self::default();
+        instance.value = MaybeUninit::new(value);
+        instance
+    }
+
+    pub fn read(mem_addr: usize) -> Self {
```

Comment:
Not completely sure whether this method should be considered safe. Leaning towards "no", but open to hearing opinions.

Reply:
Think this shouldn't be considered safe either.
It's in effect a transmutation of a [u8] into T without any check that the data validate any invariant potentially required by T. (T could be built out of garbage).

Reply:
No, this is not transmuting. This is merely copying from the flash leveraging XIP.

The transmuting will happen much later. Even calling `.value()` is not transmuting, because it's returning `MaybeUninit<T>`. The data only gets transmuted when the end-user decides to `.assume_init()` the value.

If that's the only safety-related concern, this method may be considered safe.

Reply:
I still think this needs to be unsafe, because it doesn't guarantee all safety requirements of `read_volatile`. It may, for example, be unaligned.

Reply:
@jannic Could you please explain, what could be unaligned? I don't immediately see, what that could be. The payload is aligned at 4096 bytes (or `.write()` will panic; and I should make `.read()` panic too in that case) in flash memory, which is good for any data type of up to 4096 bytes long, and larger data types are not allowed either (`::new()` will panic, etc.). And the `FlashSector<T>` is allocated on the stack according to all the rules Rust may want.

Slapping `unsafe` is simple, but if we don't know, why it's unsafe, we're just degrading the ergonomics for no reason 🤷

Reply:
Anybody can just call `FlashSector::read( FLASH_ORIGIN + 1 )` which would cause a call to `read_volatile( FLASH_ORIGIN + 1 )` which would violate the alignment requirement.

Of course, it's complete nonsense. But it would be triggering UB from safe code, so it should be prevented.

Reply:
Nobody can do that without `::read()` panicking:

https://github.com/jannic/rp2040-flash/pull/2/files#diff-bd37b4ede0f1afd74d8d015aaf023edb460bf1fd60aa68a2ce1f98d6e260712fR60

Would you prefer `::read()` to check, if the address and valid, and return a `Result<FlashSector, Error>`? Would you then consider it safe?

Reply:
I don't understand why it would panic?

Reply:
Because I perform certain assertions regarding `mem_addr`:

```rust
        assert!(mem_addr >= FLASH_ORIGIN);
        // The read address must be sector-aligned, because the write function
        // only ever allows writing at sector-aligned addresses.
        assert!(is_aligned(mem_addr, FLASH_SECTOR_SIZE));
```

I can also check if the address is within 16 MiB of the flash address space supported by RP2040.

Though, well, you could argue that the _actual_ flash size (e.g. 2 MiB on Pico) is not taken into account. In fact, this final point kinda convinces me... will be `unsafe` in a few minutes.

Reply:
While I added the checks for the upper flash address boundary, the actual flash capacity is still not checked, so I agree it's unsafe. The code is updated.

Reply:
Yes, reading from beyond the flash size is one issue.
Another point I'm not sure about is  "the memory range of the given size starting at the pointer must all be within the bounds of a single allocated object".
I have no idea what the compiler considers a "single allocated object" for direct flash access. But I wouldn't be surprised if it was allowed to use prior knowledge, for example if the memory range overlaps some static variable.

In the end, this may be uncharted territory. Eg. https://doc.rust-lang.org/reference/behavior-considered-undefined.html contains the sentence "A reference/pointer is "dangling" if it is null or not all of the bytes it points to are part of the same allocation (so in particular they all have to be part of some allocation)." One could argue that a random flash address probably is not part of any allocation, so reading it would be UB in any case. But that would also completely forbid using `volatile_read` for register access. That is obviously absurd in the context of embedded software.